pySDC/tests/test_hooks/test_log_to_file.py

```diff
@@ -15,71 +25,145 @@ def run(hook, Tend=0):
+
+    # test that we cannot overwrite if we don't want to
+    hook.allow_overwriting = False
+    with pytest.raises(FileExistsError):
+        run(**run_kwargs)
+
+    # test that we can overwrite if we do want to
+    hook.allow_overwriting = True
+    run(**run_kwargs)
+
+    # test that we cannot add solutions at times that already exist
+    hook.allow_overwriting = False
+    with pytest.raises(DataError):
+        run(**run_kwargs, t0=0.1)
 
 
 @pytest.mark.base
-def test_logging():
-    from pySDC.implementations.hooks.log_solution import LogToFile, LogSolution
+@pytest.mark.parametrize('use_pickle', [True, False])
+def test_logging(tmpdir, use_pickle, ODE=True):
+    from pySDC.implementations.hooks.log_solution import LogToPickleFile, LogSolution, LogToFile
     from pySDC.helpers.stats_helper import get_sorted
     import os
     import pickle
     import numpy as np
 
-    path = f'{os.getcwd()}/tmp'
-    LogToFile.path = path
-    Tend = 2
+    path = tmpdir
+    Tend = 0.2
+
+    if use_pickle:
+        logging_hook = LogToPickleFile
+        LogToPickleFile.path = path
+    else:
+        logging_hook = LogToFile
+        logging_hook.filename = f'{path}/FieldsIO_test.pySDC'
 
-    u0, stats = run([LogToFile, LogSolution], Tend=Tend)
+    u0, stats = run([logging_hook, LogSolution], Tend=Tend, ODE=ODE)
     u = [(0.0, u0)] + get_sorted(stats, type='u')
 
     u_file = []
     for i in range(len(u)):
-        data = LogToFile.load(i)
+        data = logging_hook.load(i)
         u_file += [(data['t'], data['u'])]
 
     for us, uf in zip(u, u_file):
-        assert us[0] == uf[0]
-        assert np.allclose(us[1], uf[1])
+        assert us[0] == uf[0], 'time does not match'
+        assert np.allclose(us[1], uf[1]), 'solution does not match'
+
+
+@pytest.mark.base
+def test_restart(tmpdir, ODE=True):
+    from pySDC.implementations.hooks.log_solution import LogSolution, LogToFile
+    import numpy as np
+
+    Tend = 0.2
+
+    # run the whole thing
+    logging_hook = LogToFile
+    logging_hook.filename = f'{tmpdir}/file.pySDC'
+
+    _, _ = run([logging_hook], Tend=Tend, ODE=ODE)
+
+    u_continuous = []
+    for i in range(20):
+        data = logging_hook.load(i)
+        u_continuous += [(data['t'], data['u'])]
+
+    # run again with a restart in the middle
+    logging_hook.filename = f'{tmpdir}/file2.pySDC'
+    _, _ = run(logging_hook, Tend=0.1, ODE=ODE)
+    _, _ = run(logging_hook, Tend=0.2, t0=0.1, ODE=ODE)
+
+    u_restart = []
+    for i in range(20):
+        data = logging_hook.load(i)
+        u_restart += [(data['t'], data['u'])]
+
+    assert np.allclose([me[0] for me in u_restart], [me[0] for me in u_continuous]), 'Times don\'t match'
+    for u1, u2 in zip(u_restart, u_continuous):
+        assert np.allclose(u1[1], u2[1]), 'solution does not match'
 
 
-if __name__ == '__main__':
-    test_logging()
+@pytest.mark.mpi4py
+@pytest.mark.mpi(ranks=[1, 4])
+def test_loggingMPI(tmpdir, comm, mpi_ranks):
```

Comment:
Why is `mpi_ranks` not used ?

Reply:
Oh, I imagine that's for the mpi test, right ? Maybe adding a bit of doc about it in the contribution guide could be good ...

Reply:
Indeed, all three arguments for this function are fixtures. `mpi_ranks` is mandatory for this version of MPI compatible pytest, see #539. If you remove it, the tests will fail and tell you to put it back in. But I can add a small comment that this is needed for the test. Although we need to take care not to copy paste this everywhere in case we don't need it anymore..
I do want to highlight how useful the `tmpdir` fixture is, though. It's perhaps not often that you need it. But if you do, it's quite handy to ensure you get a path that didn't exist before and that will not exist after the tests.